lib/language_pack/helpers/outdated_ruby_version.rb

```diff
@@ -0,0 +1,150 @@
+# Queries S3 in the background to determine
+# what versions are supported so they can be recommended
+# to the user
+#
+# Example:
+#
+#   ruby_version = LanguagePack::RubyVersion.new("ruby-2.2.5")
+#   outdated = LanguagePack::Helpers::OutdatedRubyVersion.new(
+#     current_ruby_version: ruby_version,
+#     fetcher: LanguagePack::Fetcher.new(LanguagePack::Base::VENDOR_URL, "heroku-16")
+#   )
+#
+#   outdated.call
+#   puts outdated.suggested_ruby_minor_version
+#   #=> "ruby-2.2.10"
+class LanguagePack::Helpers::OutdatedRubyVersion
+  DEFAULT_RANGE = 1..5
+  attr_reader :current_ruby_version
+
+  def initialize(current_ruby_version: , fetcher:)
+    @current_ruby_version = current_ruby_version
+    @fetcher      = fetcher
+
+    @minor_versions = [current_ruby_version]
```

Comment:
I believe we should be using [`Concurrent::Array`](http://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/Array.html) here since it's being accessed in the thread itself. Last I checked Ruby's memory model wasn't well defined, so it may not be thread safe to share memory in the thread.

Reply:
Each single operation on the array is atomic in MRI due to the GIL as long as we're not relying on something like `+=` on the shared object we're fine.

The two operations we're using are single MRI instructions:

```
@eol_versions << version
@minor_versions[i] = version
```

The first one `@eol_versions << version` is atomic due to the GIL and the order of entries doesn't matter since we later sort it.

The last one `@minor_versions[i] = version` is protected against races since each thread will have a unique integer so no two threads will try to write to the same memory.

Reply:
I still think it's better to use higher level constructs over lower level ones like `Thread`. As things change and people add new things to this, it will ensure this code continues to work.

Reply:
We actually don't have the capacity to use native ruby gems with the buildpack, we would have to vendor and maintain all of Concurrent Ruby. We've previously done this with `lpxc.rb` though this is a much smaller project.

This is what the diff would look like using concurrent Ruby:

```diff
diff --git a/lib/language_pack/helpers/outdated_ruby_version.rb b/lib/language_pack/helpers/outdated_ruby_version.rb
index 268f37b..2323e76 100644
--- a/lib/language_pack/helpers/outdated_ruby_version.rb
+++ b/lib/language_pack/helpers/outdated_ruby_version.rb
@@ -23,7 +23,7 @@ class LanguagePack::Helpers::OutdatedRubyVersion

     @minor_versions = [current_ruby_version]
     @eol_versions = []
-    @threads = []
+    @futures = []
   end

   # Enqueues checks in the background
@@ -41,7 +41,7 @@ class LanguagePack::Helpers::OutdatedRubyVersion
     return false if current_ruby_version.rbx?
     return false if current_ruby_version.jruby?

-    @threads.each(&:join)
+    @futures.each(&:value!)

     true
   end
@@ -96,7 +96,7 @@ class LanguagePack::Helpers::OutdatedRubyVersion
   # value in the series is found
   private def check_minor_versions(range: DEFAULT_RANGE, base_version: current_ruby_version, &block)
     range.each do |i|
-      @threads << Thread.new do
+      @futures << Concurrent::Future.execute do
         version = base_version.next_logical_version(i)
         next if !@fetcher.exists?("#{version}.tgz")

@@ -116,7 +116,7 @@ class LanguagePack::Helpers::OutdatedRubyVersion
   #   - 2.8.0
   private def check_eol_versions_minor(range: DEFAULT_RANGE, base_version: current_ruby_version)
     range.each do |i|
-      @threads << Thread.new do
+      @futures << Concurrent::Future.execute do
         version = base_version.next_minor_version(i)
         next if !@fetcher.exists?("#{version}.tgz")

@@ -137,7 +137,7 @@ class LanguagePack::Helpers::OutdatedRubyVersion
   #   - 3.2.0
   #   - 3.3.0
   private def check_eol_versions_major
-    @threads << Thread.new do
+    @futures << Concurrent::Future.execute do
       version = current_ruby_version.next_major_version(1)

       next if !@fetcher.exists?("#{version}.tgz")
```

The only thing that futures buy us here is that they use a thread pool rather than spawning up default threads. However I would rather not use a thread pool here, we're not using any resource limited features like database connections and the number of threads has a natural upper bound (since iteration stops when a version that does not exist is reached). I'm not sure what the default future thread pool count is but we could look it up.

When I experimented with webrick https://schneems.com/2017/08/01/is-webrick-webscale/ I saw it spike up to 101 concurrent spawned threads. 

The other thing Concurrent Ruby Future could potentially give us is if we needed to peek in on the status of an individual job i.e. `pending?`.

Reply:
Can we do a fan out/fan in strategy instead like a map/reduce? `Thread#value` returns the value of the block from the `Thread`, so we can construct the array data structure sequentially once the work is computed in parallel. While, your code actually looks ok, I think this strategy is a good pattern to reduce the cognitive overload on someone reading/touching this code in the future.

Reply:
a `Future` also doesn't required shared memory

Reply:
You can use a [SimpleExecutorService](https://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/SimpleExecutorService.html) with the `Future` if you don't want a pool

Reply:
> a Future also doesn't require shared memory

What do you mean? You mean you want the values to be stored in the future instead of a shared array? That's similar to the change Terence requested, I've updated the PR to do that instead.

> Can we do a fan out/fan in strategy instead like a map/reduce?

Done, ready for another round of reviews.